src/hotspot/share/ci/ciObjectFactory.cpp

```diff
@@ -153,9 +153,16 @@ void ciObjectFactory::init_shared_objects() {
   ciEnv::_null_object_instance = new (_arena) ciNullObject();
   init_ident_of(ciEnv::_null_object_instance);
 
-#define VM_CLASS_DEFN(name, ignore_s)                              \
-  if (vmClasses::name##_is_loaded()) \
-    ciEnv::_##name = get_metadata(vmClasses::name())->as_instance_klass();
+#define VM_CLASS_DEFN(name, ignore_s)                                \
+  if (vmClasses::name##_is_loaded()) {                               \
+    InstanceKlass* ik = vmClasses::name();                           \
+    ciEnv::_##name = get_metadata(ik)->as_instance_klass();          \
+    Array<InstanceKlass*>* interfaces = ik->transitive_interfaces(); \
```

Comment:
ciInstanceKlass objects for the vm classes all need to be allocated from the same long lived arena that's created in ciObjectFactory::initialize() because they are shared between compilations. Without that code, the ciInstanceKlass for a particular vm class is in the long lived arena but the ciInstanceKlass objects for the interfaces are created later when they are needed in the arena of some compilation. Once that compilation is over the interface objects are destroyed but still referenced from shared types such as TypeInstPtr::MIRROR.